### PR TITLE
Add govulncheck presubmit to expose go vulnerabilities in a PR

### PR DESCRIPTION
**Pull request by ArkaSaha30**
This PR will add a govulncheck presubmit to check for go vulnerabilities when a PR  is opened for go module changes. It will trigger the script from ~~`sig-security/sig-security-tooling/govulncheck/hack/govulncheck-presubmit.sh`~~ `kubernetes/hack/verify-govulncheck.sh`.

Fixes: https://github.com/kubernetes/sig-security/issues/99
Parent Issue: https://github.com/kubernetes/sig-security/issues/95

### CODE REVIEW

**Comment by rjsadow** [comment]
I don't see any issues with this @ArkaSaha30. Would you be able to run this on the EKS cluster instead of the default cluster like it's currently set up? You can use https://github.com/kubernetes/test-infra/pull/30556/files as a guide on what updates are needed.

File: config/jobs/kubernetes/sig-security/govulncheck-presubmit.yaml [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    optional: true
+    run_if_changed: '^(go.mod|go.sum|vendor)'
+    spec:
+      containers:
+      - image: golang
+        command:
+        - /bin/bash
+        args:
+        - -c
+        - |
```
Comment (rjsadow) [inline config/jobs/kubernetes/sig-security/govulncheck-presubmit.yaml:23]: I think this pipe here as is will cause issues. The pipe in yaml indicates a multi-line value. However following this is the script value in a separate array element. See https://yaml.org/spec/1.2-old/spec.html#id2795688. 

I'd recommend either:

- Removing this line

or
-  remove the quotes and dash from the following line. i.e.
```
  args:
  - -c
  - |
    cd sig-security-tooling/govulncheck/hack/ && ./govulncheck-presubmit.sh
```

Comment (ArkaSaha30) [reply config/jobs/kubernetes/sig-security/govulncheck-presubmit.yaml:23]: Sure, thank you 👍

**Comment by rjsadow** [comment]
Will this job just create the vulnerability artifact that can be reviewed for informational purposes, or is the intent to have some kind of gatekeeping element to fail if (for example) there are any CVEs found? If it's for gatekeeping, I'm not currently understanding how/where the job will fail out of the test.

**Comment by PushkarJ** [comment]
> Will this job just create the vulnerability artifact that can be reviewed for informational purposes, or is the intent to have some kind of gatekeeping element to fail if (for example) there are any CVEs found? If it's for gatekeeping, I'm not currently understanding how/where the job will fail out of the test.

This will be mostly for informational purposes (atleast for now) since we want the job to not block a PR right now. This will allow us to observe and iterate over the accuracy of the scan results and then potentially in future act as gatekeeping PR if needed

**Comment by rjsadow** [comment]
Awesome, thank you @ArkaSaha30 and @PushkarJ 
/lgtm

**Comment by ArkaSaha30** [comment]
Made a small change to the presubmit after testing locally:
```
-          cd sig-security-tooling/govulncheck/hack/ && ./govulncheck-presubmit.sh
+          cd ../kubernetes
+          ../sig-security/sig-security-tooling/govulncheck/hack/govulncheck-presubmit.sh
```
The `govulncheck-presubmit.sh` needs to be triggered from inside `kubernetes` directory.

File: config/jobs/kubernetes/sig-security/govulncheck-presubmit.yaml [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+          cd ../kubernetes
```
Comment (rjsadow) [inline config/jobs/kubernetes/sig-security/govulncheck-presubmit.yaml:25]: I believe you can get rid of this line if you set `k/k` in `extra_refs`. i.e.

```
      extra_refs:
      - org: kubernetes
        repo: kubernetes
        base_ref: main
        workdir: true
      - org: kubernetes
        repo: sig-security
        base_ref: main

```

Comment (ArkaSaha30) [reply config/jobs/kubernetes/sig-security/govulncheck-presubmit.yaml:25]: But this is a presubmit...by adding `kubernetes` in `extra_refs` will it not overwrite the PR specific clone of `k/k`?

Comment (rjsadow) [reply config/jobs/kubernetes/sig-security/govulncheck-presubmit.yaml:25]: gah, yes. You're right :)

**Review by rjsadow — APPROVED** [review]

**Comment by BenTheElder** [comment]
https://github.com/kubernetes/sig-security/pull/101#pullrequestreview-1614344072
/hold

**Comment by ArkaSaha30** [comment]
Pushing the changes as per discussion threads:
- https://github.com/kubernetes/sig-security/pull/101#pullrequestreview-1614344072
- https://kubernetes.slack.com/archives/CHGFYJVAN/p1694040913317869

**Comment by rjsadow** [comment]
@ArkaSaha30 @PushkarJ are we ok to close this based on https://github.com/kubernetes/kubernetes/pull/120562#issuecomment-1722098423 or do you see any other uses for having a separate job?

**Comment by liggitt** [comment]
The default verify script will make sure the PR doesn't make things *worse*, but I think we'd want a periodic job to check if existing state has issues, right?

**Comment by PushkarJ** [comment]
@liggitt yes we need the periodic one for sure. We are tracking it separately with this issue https://github.com/kubernetes/sig-security/issues/100

**Comment by rjsadow** [comment]
I'm going to close this pr for now since https://github.com/kubernetes/kubernetes/pull/120562 will add it to `k/k` presubmits and there's a separate issue tracking the periodic. 

/close

**Pull request closed without merging**